enhancements/authentication/login-logout-events.md

```diff
@@ -0,0 +1,335 @@
+type OAuthSpec struct {
+	// identityProviders is an ordered list of ways for a user to identify themselves.
+	// When this list is empty, no identities are provisioned for users.
+	// +optional
+	// +listType=atomic
+	IdentityProviders []IdentityProvider `json:"identityProviders,omitempty"`
+
+	// tokenConfig contains options for authorization and access tokens
+	TokenConfig TokenConfig `json:"tokenConfig"`
+
+	// templates allow you to customize pages like the login page.
+	// +optional
+	Templates OAuthTemplates `json:"templates"`
+	// audit specifies what should be audited in the context of OAuthServer.
+	// +optional
+	// +kubebuilder:default:={"profile":"WriteLoginEvents"}
+	Audit OAuthAudit `json:"audit"`
+}
+
+// OAuthAudit specifies the Audit profile in use.
+type OAuthAudit struct {
+	// profile is a simple drop in profile type that can be turned off by
+	// setting it to "None" or it can be turned on by setting it to
+	// "WriteLoginEvents".
+	// +kubebuilder:default:="WriteLoginEvents"
+	Profile OAuthAuditProfileType `json:"profile,omitempty"`
+}
+
+// OAuthAuditProfileType defines a simple audit profile, which can turn OAuth
+// authentication audit logging on or off.
+// +kubebuilder:validation:Enum=None;WriteLoginEvents
+type OAuthAuditProfileType string
+
+const (
+	// "None" disables audit logs.
+	OAuthNoneAuditProfileType AuditProfileType = "None"
+
+	// "WriteLoginEvents" logs login and login failure events.
+	// This is the default.
+	OAuthWriteLoginEventsProfileType AuditProfileType = "WriteLoginEvents"
+)
+```
+
+### User Stories
+
+#### As an administrator I want to inspect successful login and login failure attempts
+
+Use `oc adm must-gather -- /usr/bin/gather_audit_logs` and inspect oauth-server audit logs.
+
+#### As an administrator I want to inspect logout events and the deleted user
+
+Use `oc adm must-gather -- /usr/bin/gather_audit_logs` and inspect apiserver audit logs for deleted `oauthaccesstoken` resources.
+Audit log entries include the deleted user.
+
+#### As an administrator I want to inspect the source IPs of the client triggering an authentication event
+
+Use `oc adm must-gather -- /usr/bin/gather_audit_logs` and inspect oauth-server audit logs for auth events and inspect the IP within the event.
+
+### API Extensions
+
+We introduce a new property called `audit` to the `OAuth` CRD.
+
+This enhancement is not expected to affect any other team or compoment outside of auth.
+
+### Operational Aspects of API Extensions
+
+There is no explicit operational involvement expected. The property is set by default and is optional.
+
+It could be turned off by operational interaction with the API extension.
+
+#### Failure Modes
+
+N/A
+
+#### Support Procedures
+
+If the `audit` can't be set in the CR, support would need to file a bugzilla.
+If there are no audit events for login, login failures or logout events, verify that the `audit` values is set in the `oauth` CR.
```

Comment:
The `audit` property does not influence logging of logout events

Reply:
It doesn't? What is then the whole purpose of setting `audit`?

Will the `audit` property not tell the `cluster-authentication-operator` to enable audit logging on `oauth-server`?

Reply:
As far as I know, it only influences login events, and also logs some login failures. Hence the `WriteLoginEvents` value that you can configure for the profile.

Reply:
But isn't the goal to turn the audit logging on **and then** add the logging of the other events too?

So there will be a certain moment in time, before we add more logging events to the `oauth-server`, where we will only log a very limited set of events. But once we adjusted `oauth-server`, it will log login, login failures and logout... oh noes... you are right. That is true. The logout events are logged by extending what is being logged on logout and not in particular by the profile. So we have different technical paths to solve the goal.

Thanks a lot for the insight and thorough review :D This enhancement helped me to gain so much more clarity on that topic.

Reply:
@ibihim yes, logout events cannot be captured in oauth-server

Reply:
@ibihim as described in this OEP, logout events are captured only via deletion of oauthaccesstokens, subject auf audit policy configuration in oauth-apiserver.